OneCommunityGlobal/HGNRest
# Diego enforcing unit tests in the backend

<img width="774" alt="image" src="https://github.com/OneCommunityGlobal/HGNRest/assets/26840798/d94b18fe-1337-4756-ab21-1cd8d90c0bb8">

# Description
Adds jest, lint-staged and husky to run unit tests on staged files before being committed. Currently, the backend doesn't have any unit tests and we need to enforce running unit tests before committing our code to help reducing errors in production.

## Related PRS:
N/A

## Main changes explained:
- Runs unit tests on staged files
- Stops committing phase when tests fail

## How to test:
1. check into current branch
2. do `npm install`
3. checkout into a new branch of your own to test this functionality (since we will create a new commit to test this). eg. git checkout -b yourname/test_jest
4. if running on Mac or Linux run: `npm run prepare` (this will make husky files executable)
5. add a new file. e.g. math.js
6. add some logic for testing. e.g.:
`const sum = (a, b) => a + b;
const substract = (a, b) => a - b;
module.exports = { sum, substract };`
7. add a test file: e.g. `math.spec.js`:
```
describe('math module', () => {
  test('ensure a + b returns the correct value', () => {
    expect(sum(1, 2)).toBe(3);
  });

  test('ensure a - b returns the correct value', () => {
    expect(substract(1, 2)).toBe(-1);
  });
});
```
8. enter the following command: npm run test (it should run all of your test suites)
9. enter the following command: npm run test:unit (it should run all the test suites on `**.spec.js` files on watch mode)
10. while running npm run test:unit, make changes to your test files and save them, the jest watcher will run the unit tests again for every file save.
11. you may add two additional dummy files for testing another piece of logic.
12. add only the `math.js` and `math.spec.js` or whatever names you used for your production and test files.
13. commit your changes (unit tests should be run only for staged files)

## Screenshots or videos of changes:
https://github.com/OneCommunityGlobal/HGNRest/assets/26840798/0480d3a7-ef82-4d3c-8761-3714d8a8d60b

## Note:
Unit tests will have to be added to the existing code base and it will take a few weeks at least to cover all of the files involved in making API routes to work (we need some volunteers to help us cover existing codebase with tests). Once we have covered all of these files, we will be able to start working with Test Driven Development: [part 1](https://www.dropbox.com/scl/fi/i5cmjqw6zgsrztes8n1za/tdd-development-process-part-1.mp4?rlkey=g36omvw72vktoki8nwpujuc97&dl=0), [part 2](https://www.dropbox.com/scl/fi/fyqrvtl5tclfo7t2viged/tdd-development-process-part-2.mp4?rlkey=sywsfb8sal08b83bgtbtyvyel&dl=0). Please read the [documentation](https://docs.google.com/document/d/1s-AJrn4GgcH0UJnFOw64w6ZmfKlmOAPlMZZD51sV_80/edit) on it).

Finally, once we have tested all the files involved in making API routes to work, we will be able to enforce a code coverage threshold which will help us enforce high standards when developing: [video](https://www.dropbox.com/scl/fi/9gyd9b791zwou9xcr92h2/Test-coverage-pre-push.mp4?rlkey=lol2s4zvrp07hxgt4046rroaw&dl=0)
